Fix Button Enabling and Data Clearing Behavior for Offline Vehicle State

smadi0x86 opened the pull request:
# Fix Button Enabling and Data Clearing Behavior for Offline Vehicle State

Description
-----------
**This update addresses two key issues related to the data clearing functionality within the application:**

**1. Button Enabling:** The "Clear" button is now enabled even when the vehicle is offline. Previously, the button was only enabled when the vehicle was online and not syncing, which prevented users from clearing data when the vehicle was offline.

**2. Data Clearing Behavior:** Adjustments were made to ensure that the removeAllFromVehicle() functions in various controllers (Mission, GeoFence, RallyPoint) are executed even if the vehicle is offline. This change allows the application to clear all relevant data regardless of the vehicle’s online status. Warnings are logged when these actions are attempted while offline, but the operations will still proceed.

**Note:** This PR is still not ready, this was created based on the issue author request for potential collaboration.

After these changes, clearing isn't working when vehicle is offline and I am getting this error even after clicking on `yes`:
```
qml: User cancelled: Did not clear mission items from the vehicle 
```
Test Steps
-----------

**1. Verify Button Functionality:**

- Launch the application and ensure the vehicle is offline.
- Check if the "Clear" button is enabled in the user interface.
- Click the "Clear" button and confirm that the appropriate data clearing actions are attempted.

**2. Validate Data Clearing:**

- When the vehicle is offline, trigger data clearing operations for missions, geofences, and rally points.
- Verify that the operations proceed and that warnings are logged appropriately, indicating the offline status but still performing the clear actions.

**3. Check Online Behavior:**

- Switch the vehicle to an online state.
- Ensure the "Clear" button is still enabled and functions correctly.
- Confirm that data clearing operations proceed as expected when the vehicle is online.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] [Review Contribution Guidelines](https://github.com/mavlink/qgroundcontrol/blob/master/.github/CONTRIBUTING.md).
- [x] [Review Code of Conduct](https://github.com/mavlink/qgroundcontrol/blob/master/.github/CODE_OF_CONDUCT.md).
- [ ] I have tested my changes.

Related Issue
-----------
https://github.com/mavlink/qgroundcontrol/issues/11544


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

DonLakeFlyer commented:
There are a number of problems here, some I discussed in #11544:

As the UI currently stands the "Clear" button is associated with a vehicle:
<img width="272" alt="Screenshot 2024-08-01 at 9 36 44 AM" src="https://github.com/user-attachments/assets/ae924a1e-85da-4b6b-9e05-6464b82873cf">
So it doesn't make sense for things under "Vehicle" to work when there is no Vehicle. You have also modified methods called removeAllFromVehicle to not do what they say they do. They are no longer associated with a vehicle.

I think it's a bad idea to allow removeAllFromVehicle while a sync is on progress. There is a sync state machine running which you then intersperse another mission command on top off while the sync continues to run.  No idea what sort of strange problems this could lead to. Also don't understand why this is something interesting to support. In order to support well there would need to be the ability to cancel a sync cleanly and then do the remove all.

As I said in the Issue the ability to start over from scratch when not connected to a vehicle is already supported by the  Create a Blank Plan button. I believe that is the exact functionality you want, it's just not called Clear. I realize this may not be as obvious as a "Clear" button but I'm also not quite sure how confusing it is for most people. In order to resolve this it would require rework of the overall UI, not just enabling things that weren't enabled before. At this point in looking at it I'm not super convinced this is a big problem that needs solving. Let me ask the community what they think.

smadi0x86 commented:
> There are a number of problems here, some I discussed in #11544:
> 
> As the UI currently stands the "Clear" button is associated with a vehicle: <img alt="Screenshot 2024-08-01 at 9 36 44 AM" width="272" src="https://private-user-images.githubusercontent.com/5876851/354307364-ae924a1e-85da-4b6b-9e05-6464b82873cf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1NTEwNjEsIm5iZiI6MTcyMjU1MDc2MSwicGF0aCI6Ii81ODc2ODUxLzM1NDMwNzM2NC1hZTkyNGExZS04NWRhLTRiNmItOWUwNS02NDY0YjgyODczY2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDFUMjIxOTIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2YxMDczYTM5MjYyMGI2ZDlhYzgzODhkZWI0YjljNjVkYTYwZTEyNjk3MjA2YTczYTY1YzllODJiMjNlYmMzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5iK7Gvp0weZDN6s_YWl9rsotWrpldY77ncIiHLmlcBI"> So it doesn't make sense for things under "Vehicle" to work when there is no Vehicle. You have also modified methods called removeAllFromVehicle to not do what they say they do. They are no longer associated with a vehicle.
> 
> I think it's a bad idea to allow removeAllFromVehicle while a sync is on progress. There is a sync state machine running which you then intersperse another mission command on top off while the sync continues to run. No idea what sort of strange problems this could lead to. Also don't understand why this is something interesting to support. In order to support well there would need to be the ability to cancel a sync cleanly and then do the remove all.
> 
> As I said in the Issue the ability to start over from scratch when not connected to a vehicle is already supported by the Create a Blank Plan button. I believe that is the exact functionality you want, it's just not called Clear. I realize this may not be as obvious as a "Clear" button but I'm also not quite sure how confusing it is for most people. In order to resolve this it would require rework of the overall UI, not just enabling things that weren't enabled before. At this point in looking at it I'm not super convinced this is a big problem that needs solving. Let me ask the community what they think.

Yeah as mentioned in the issue this requires a community decision, from my side I changed this PR to a draft until further notice. And yes at this point its a re-design to the UI not only to fix the clear view.

DonLakeFlyer commented:
I'm closing this since it is incorrect. I've submitted another pull that should make things a little better.
